packages/flutter/lib/src/cupertino/sheet.dart

```diff
@@ -0,0 +1,349 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
+
+import 'colors.dart';
+import 'route.dart';
+
+// Offset from offscreen below to stopping below the top of the screen.
+final Animatable<Offset> _kBottomUpTween = Tween<Offset>(
+  begin: const Offset(0.0, 1.0),
+  end: const Offset(0.0, 0.08),
+);
+
+final Animatable<Offset> _kFullBottomUpTween = Tween<Offset>(
+  begin: const Offset(0.0, 1.0),
+  end: const Offset(0.0, -0.02),
+);
+
+// Offset slightly up when a sheet gets covered by another sheet.
+final Animatable<Offset> _kMidUpTween = Tween<Offset>(
+  begin: Offset.zero,
+  end: const Offset(0.0, -0.005),
+);
+
+// Offset from top of screen to slightly down when covered by a sheet.
+final Animatable<Offset> _kTopDownTween = Tween<Offset>(
+  begin: Offset.zero,
+  end: const Offset(0.0, 0.07),
+);
+
+// Amount the sheet in the background scales down. Found by measuring the width
+// of the sheet in the background and comparing against the screen width. The
+// scale transition will go from a default of 1.0 to 1.0 - _kSheetScaleFactor.
+const double _kSheetScaleFactor = 0.0835;
+
+final Animatable<double> _kScaleTween = Tween<double>(begin: 1.0, end: 1.0 - _kSheetScaleFactor);
+
+/// Docs placeholder
+Future<T?> showCupertinoSheet<T>({
+  required BuildContext context,
+  required WidgetBuilder pageBuilder,
+  bool useRootNavigator = true,
+  bool useNestedNavigation = false,
+}) {
+  final BuildContext topLevelSheetContext = CupertinoSheetController.maybeOf(context)?.topLevelContext ?? context;
+
+  final WidgetBuilder builder = switch (useNestedNavigation) {
+    false => pageBuilder,
+    true => (BuildContext context) {
+      return Navigator(
+        initialRoute: '/',
+        onGenerateRoute: (RouteSettings settings) {
+          return CupertinoPageRoute<void>(
+            builder: (BuildContext context) {
+              return PopScope(
```

Comment:
Consider trying this out with
 * go_router
 * deep linking, and/or specific web-style URL requirements.

I wonder what kind of requirements you'll run into to get those working. Maybe you will need some kind of builder parameter instead of just the plain boolean.